### ISSUE TITLE
"Unable to load shared library 'MonoPosixHelper' or one of its dependencies" on ARM64 Alpine Linux

### ISSUE THREAD

**Issue by romner-set** [slskd/slskd#1013]
I'm running into this error trying to setup a non-docker relay controller on a free Oracle ARM VM running Alpine. I'm using the [slskd-0.19.1-linux-musl-arm64.zip](https://github.com/slskd/slskd/releases/download/0.19.1/slskd-0.19.1-linux-musl-arm64.zip) binary release, this is the entire error message:
```
[01:06:58 INF] Configuration complete.  Starting application...
[01:06:59 FTL] Application terminated unexpectedly
System.DllNotFoundException: Unable to load shared library 'MonoPosixHelper' or one of its dependencies. In order to help diagnose loading problems, consider using a tool like strace. If you're using glibc, consider setting the LD_DEBUG environment variable:
Error loading shared library /root/.net/slskd/6zwXsAVSvoI0dYsMXU8XYITRGqrN1KA=/MonoPosixHelper.so: No such file or directory
Error loading shared library /etc/slskd/MonoPosixHelper.so: No such file or directory
Error loading shared library MonoPosixHelper.so: No such file or directory
Error loading shared library ld-linux-aarch64.so.1: No such file or directory (needed by /root/.net/slskd/6zwXsAVSvoI0dYsMXU8XYITRGqrN1KA=/libMonoPosixHelper.so)
Error loading shared library /etc/slskd/libMonoPosixHelper.so: No such file or directory
Error loading shared library libMonoPosixHelper.so: No such file or directory
Error loading shared library /root/.net/slskd/6zwXsAVSvoI0dYsMXU8XYITRGqrN1KA=/MonoPosixHelper: No such file or directory
Error loading shared library /etc/slskd/MonoPosixHelper: No such file or directory
Error loading shared library MonoPosixHelper: No such file or directory
Error loading shared library /root/.net/slskd/6zwXsAVSvoI0dYsMXU8XYITRGqrN1KA=/libMonoPosixHelper: No such file or directory
Error loading shared library /etc/slskd/libMonoPosixHelper: No such file or directory
Error loading shared library libMonoPosixHelper: No such file or directory

   at Mono.Unix.Native.Syscall.sysconf(SysconfName name, Errno defaultError)
   at Mono.Unix.Native.Syscall.sysconf(SysconfName name)
   at Prometheus.SystemMetrics.Collectors.LinuxCpuUsageCollector.CreateMetrics(MetricFactory factory)
   at Prometheus.SystemMetrics.SystemMetricsHostedService.StartAsync(CancellationToken cancellationToken)
   at Microsoft.Extensions.Hosting.Internal.Host.StartAsync(CancellationToken cancellationToken)
   at Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.RunAsync(IHost host, CancellationToken token)
   at Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.RunAsync(IHost host, CancellationToken token)
   at Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.Run(IHost host)
   at slskd.Program.Main(String[] args)
```

The binary seems to be linked correctly, comparing it to a system binary:
```bash
$ file /bin/busybox
/bin/busybox: ELF 64-bit LSB pie executable, ARM aarch64, version 1 (SYSV), dynamically linked, interpreter /lib/ld-musl-aarch64.so.1, BuildID[sha1]=044386987bb479314c5fb9741167734e720bcc3e, stripped

$ file ./slskd
./slskd: ELF 64-bit LSB shared object, ARM aarch64, version 1 (SYSV), dynamically linked, interpreter /lib/ld-musl-aarch64.so.1, BuildID[sha1]=9ba6f2b0097c07f697a9ad38d106290d5b535f8b, not stripped
```

**Comment by jpdillingham** [comment]
For Alpine you'll want to make sure you have all of the necessary runtime dependencies installed.  Here's the official list Microsoft uses for their `runtime-deps` container: https://github.com/dotnet/dotnet-docker/blob/main/src/runtime-deps/6.0/alpine3.18/amd64/Dockerfile

You may have already done this, as I think typically you'll get cryptic 'file not found' messages when starting the app, and this issue is arising later when the metrics stuff is being bootstrapped.  Here's a corresponding issue: https://github.com/Daniel15/prometheus-net.SystemMetrics/issues/6

I can add a configuration flag that prevents the metrics causing this from loading, which will either solve the problem or move it to the next thing that depends on the missing library.

I will see if I can reproduce this.  If you wouldn't mind, could you try to install the dependencies from the Dockerfile I linked and see if that solves anything?

**Comment by romner-set** [comment]
> For Alpine you'll want to make sure you have all of the necessary runtime dependencies installed. Here's the official list Microsoft uses for their `runtime-deps` container: https://github.com/dotnet/dotnet-docker/blob/main/src/runtime-deps/6.0/alpine3.18/amd64/Dockerfile
> 
> You may have already done this, as I think typically you'll get cryptic 'file not found' messages when starting the app, and this issue is arising later when the metrics stuff is being bootstrapped. Here's a corresponding issue: [Daniel15/prometheus-net.SystemMetrics#6](https://github.com/Daniel15/prometheus-net.SystemMetrics/issues/6)
> 
> I can add a configuration flag that prevents the metrics causing this from loading, which will either solve the problem or move it to the next thing that depends on the missing library.
> 
> I will see if I can reproduce this. If you wouldn't mind, could you try to install the dependencies from the Dockerfile I linked and see if that solves anything?

Sure, just downloaded all of them (krb5-conf, libcom_err, libverto and krb5-libs were missing) and I'm still getting the exact same error. Let me know if there's anything else you'd want me to test.

**Comment by Daniel15** [comment]
> Here's a corresponding issue: https://github.com/Daniel15/prometheus-net.SystemMetrics/issues/6

Sorry about that! I suspect Mono don't have an ARM64 build of `libMonoPosixHelper`. I'll see if I can remove the Mono.Posix library and just PInvoke `sysconf` directly.

**Comment by Daniel15** [comment]
@jpdillingham @romner-set I just published v3.1.0 of the `prometheus-net.SystemMetrics` library, which removes the dependency on `Mono.Posix`. This *should* fix the issue with ARM64, however currently I don't have any ARM64 systems to test with.

**Comment by jpdillingham** [comment]
@Daniel15 thanks! I'll get this out in the next build and do some testing.
